c10/mobile/CPUProfilingAllocator.cpp

```diff
@@ -49,6 +31,36 @@ struct MemEvent {
     time(t), allocation_id(id), size(s), type(e) {}
 };
 
+bool validate_allocation_plan(
+    const std::vector<MemEvent>& alloc_events,
+    const std::vector<uint64_t>& allocation_offsets) {
+  std::set<MemBlock> allocations;
+  for (const auto& event : alloc_events) {
+    auto alloc_id = event.allocation_id;
+    // Skip allocations not managed by AllocationPlan
+    if (allocation_offsets[alloc_id] == std::numeric_limits<uint64_t>::max()) {
+      continue;
+    }
+    auto start_offset = allocation_offsets[alloc_id];
+    auto end_offset = allocation_offsets[alloc_id] + event.size;
+    if (event.type == EventType::Allocate) {
+      if (!allocations.emplace(start_offset, end_offset).second) {
```

Comment:
if you really want to validate - you should check that the inserted block doesn't overlap with existing blocks here

Reply:
@dzhulgakov Thats what this emplace is supposed to do. Did I miss something?

Reply:
Emplace will only complain if the object is equal (which in this case is probably expressed via `!(a<b) && !(b<a)`). I originally assumed that you just sort the items in the set using their `start_offsets`, but it looks like you actually define the partial order above: `    return end_offset <= other.start_offset;	`.

I think it's technically not correct to create `std::set` with such comparison operator. STL documentation requires "strict weak ordering" (http://www.cplusplus.com/reference/set/set/), i.e. https://en.wikipedia.org/wiki/Weak_ordering#Strict_weak_orderings . In particular string weak ordering requires "transitivity of incompatibility":

> For all x, y, z in S, if x is incomparable with y (neither x < y nor y < x hold), and y is incomparable with z, then x is incomparable with z (transitivity of incomparability).

However, you operator doesn't satisfy it, it's a non-strict weak ordering, e.g.:

```
x = (1, 10)
y = (5, 15)
z = (11, 20)
x < y || y < x == False # x and y are incompatible
y < z || z < y == False # y and z are incompatible
x < z # boom!
```

What it means that under certain overlapping circumstances you can get crash/assertion/runtime error being thrown from std::set operation.

I'd also give you another reason to not do it: it's very non-obvious to rely on such fancy comparison operator and it's very likely the reader of this code will overlook it (like I did).

To summarize: please change comparison operator to just look at starts only (so that it's a strict ordering) and then add necessary checks here that the inserted element doesn't overlap with the existing ones. It should be sufficient to just check the iterator just after the inserted one and verify that its start is after the inserted end.

Reply:
@dzhulgakov, thanks for pointing this out. I did miss "strict weak ordering" part and thanks for the explanation. That does make sense. Also for readability part, I do agree. It does get a bit cryptic. I should have put comments there at the least.

I will make changes.

Thanks once again. This was informative.